File: app/src/main/java/com/sdwfqin/quickseed/ui/mvvm/WeatherViewModel.java
package com.sdwfqin.quickseed.ui.mvvm;

import androidx.lifecycle.MutableLiveData;

import com.sdwfqin.quicklib.mvvm.BaseViewModel;
import com.sdwfqin.quickseed.retrofit.RetrofitClient;

import java.util.HashMap;
import java.util.Map;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

/**
 * 天气ViewModel
 * <p>
 * 注意：ViewModel 绝不能引用视图、Lifecycle 或可能存储对 Activity 上下文的引用的任何类。
 *
 * @author 张钦
 * @date 2020/4/15
 */
public class WeatherViewModel extends BaseViewModel {

    public final MutableLiveData<WeatherBean> weatherBean = new MutableLiveData<>();

    public void loadWeather() {
        isLoading.postValue(true);
        Map<String, Object> map = new HashMap();
        RetrofitClient
                .getInstance()
                .gService.getWeather(map)
                .enqueue(new Callback<WeatherBean>() {
                    @Override
                    public void onResponse(Call<WeatherBean> call, Response<WeatherBean> response) {
                        weatherBean.postValue(response.body());
                        isLoading.postValue(false);
                    }

                    @Override
                    public void onFailure(Call<WeatherBean> call, Throwable t) {
                        isLoading.postValue(false);
                        networkError.postValue(t);
                    }
                });
    }
}

File: app-kt/src/main/java/io/github/sdwfqin/app_kt/ui/mvvm/WeatherViewModel.kt
package io.github.sdwfqin.app_kt.ui.mvvm

import androidx.hilt.lifecycle.ViewModelInject
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import com.blankj.utilcode.util.LogUtils
import com.sdwfqin.quicklib.mvvm.BaseViewModel
import io.github.sdwfqin.app_kt.data.bean.WeatherBean
import io.github.sdwfqin.app_kt.data.repository.WeatherRepository
import kotlinx.coroutines.launch
import java.util.*

/**
 * 天气ViewModel
 *
 * @author 张钦
 * @date 2020/4/15
 */
class WeatherViewModel @ViewModelInject constructor(
        private val repository: WeatherRepository
) : BaseViewModel() {

    val weatherBean = MutableLiveData<WeatherBean>()

    fun loadWeather() {

        isLoading.postValue(true)

        val map: Map<String, Any> = HashMap<String, Any>()

        launch({
            weatherBean.postValue(repository.getWeather(map))
        }, {
            LogUtils.e(it)
            networkError.postValue(it)
        }, {
            isLoading.postValue(false)
        })
    }

    private fun launch(block: suspend () -> Unit, error: suspend (Throwable) -> Unit, complete: suspend () -> Unit) = viewModelScope.launch {
        try {
            block()
        } catch (e: Throwable) {
            error(e)
        } finally {
            complete()
        }
    }
}